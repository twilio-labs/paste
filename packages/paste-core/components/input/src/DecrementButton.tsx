import * as React from 'react';
import type {BoxProps} from '@twilio-paste/box';
import {Button} from '@twilio-paste/button';
import {ChevronDownIcon} from '@twilio-paste/icons/esm/ChevronDownIcon';

export interface DecrementButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  i18nStepDownLabel?: string;
  element?: BoxProps['element'];
}

export const DecrementButton = React.forwardRef<HTMLButtonElement, DecrementButtonProps>(
  ({i18nStepDownLabel = 'step value down', element, ...props}, ref) => {
    return (
      <Button
        {...props}
        ref={ref}
        element={`${element}_DECREMENT_BUTTON`}
        variant="reset"
        size="reset"
        type="button"
        // @ts-expect-error remove when Reset Button types extends BoxProps
        borderRadius="borderRadius20"
        backgroundColor="colorBackground"
        marginRight="space30"
      >
        <ChevronDownIcon
          decorative={false}
          title={i18nStepDownLabel}
          size="sizeIcon05"
          element={`${element}_DECREMENT_ICON`}
        />
      </Button>
    );
  }
);
DecrementButton.displayName = 'Decrement';
