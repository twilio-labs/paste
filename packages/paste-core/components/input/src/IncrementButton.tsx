import * as React from 'react';
import type {BoxProps} from '@twilio-paste/box';
import {Button} from '@twilio-paste/button';
import {ChevronUpIcon} from '@twilio-paste/icons/esm/ChevronUpIcon';
import type {HTMLPasteProps} from '@twilio-paste/types';

export interface IncrementButtonProps extends HTMLPasteProps<'button'> {
  i18nStepUpLabel?: string;
  element?: BoxProps['element'];
}

export const IncrementButton = React.forwardRef<HTMLButtonElement, IncrementButtonProps>(
  ({i18nStepUpLabel = 'step value up', element, ...props}, ref) => {
    return (
      <Button
        {...props}
        ref={ref}
        element={`${element}_INCREMENT_BUTTON`}
        variant="reset"
        size="reset"
        type="button"
        // @ts-expect-error remove when Reset Button types extends BoxProps
        borderRadius="borderRadius20"
        backgroundColor="colorBackground"
        marginRight="space30"
      >
        <ChevronUpIcon
          decorative={false}
          title={i18nStepUpLabel}
          size="sizeIcon05"
          element={`${element}_INCREMENT_ICON`}
        />
      </Button>
    );
  }
);
IncrementButton.displayName = 'Increment';
