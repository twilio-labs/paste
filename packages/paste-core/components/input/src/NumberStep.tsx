import * as React from 'react';
import type {BoxProps} from '@twilio-paste/box';
import {Button} from '@twilio-paste/button';
import {ChevronUpIcon} from '@twilio-paste/icons/esm/ChevronUpIcon';
import {ChevronDownIcon} from '@twilio-paste/icons/esm/ChevronDownIcon';

export interface IncrementButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  i18nStepUpLabel?: string;
  element?: BoxProps['element'];
}
export interface DecrementButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  i18nStepDownLabel?: string;
  element?: BoxProps['element'];
}

export const IncrementButton = React.forwardRef<HTMLButtonElement, IncrementButtonProps>(
  ({i18nStepUpLabel = 'step value up', element, ...props}, ref) => {
    return (
      <Button
        {...props}
        ref={ref}
        element={`${element}_INCREMENT_BUTTON`}
        variant="reset"
        size="reset"
        type="button"
        // @ts-expect-error remove when Reset Button types extends BoxProps
        borderRadius="borderRadius20"
        backgroundColor="colorBackground"
        marginRight="space30"
      >
        <ChevronUpIcon
          decorative={false}
          title={i18nStepUpLabel}
          size="sizeIcon05"
          element={`${element}_INCREMENT_ICON`}
        />
      </Button>
    );
  }
);
IncrementButton.displayName = 'Increment';

export const DecrementButton = React.forwardRef<HTMLButtonElement, DecrementButtonProps>(
  ({i18nStepDownLabel = 'step value down', element, ...props}, ref) => {
    return (
      <Button
        {...props}
        ref={ref}
        element={`${element}_DECREMENT_BUTTON`}
        variant="reset"
        size="reset"
        type="button"
        // @ts-expect-error remove when Reset Button types extends BoxProps
        borderRadius="borderRadius20"
        backgroundColor="colorBackground"
        marginRight="space30"
      >
        <ChevronDownIcon
          decorative={false}
          title={i18nStepDownLabel}
          size="sizeIcon05"
          element={`${element}_DECREMENT_ICON`}
        />
      </Button>
    );
  }
);
DecrementButton.displayName = 'Decrement';
