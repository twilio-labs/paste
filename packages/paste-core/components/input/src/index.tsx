export { Input, InputElement } from "./Input";
export type { InputProps, InputVariants } from "./Input";
