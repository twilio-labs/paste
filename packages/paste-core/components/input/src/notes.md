TO DO

- BUG: step=3 value starts at 0 and first step goes to 1 then starts stepping correctly
- tests and more stories (controlled & uncontrolled)
- MONDAY: ask about hover/focus styles on reset button, add to ticket
  - if so, add to ticket AC to adjust hover/focus styles on number nudger buttons too or make another ticket dependent on the button one
