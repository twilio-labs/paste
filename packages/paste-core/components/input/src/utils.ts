export const PROPS_TO_BLOCK: string[] = ['className', 'style', 'size', 'height', 'width'];

type Props = {[key: string]: any};

export const safelySpreadFormControlProps = (props: Props): Props => {
  // https://www.measurethat.net/Benchmarks/Show/6642/0/for-in-vs-reduce-vs-pick#latest_results_block
  return Object.keys(props).reduce((newProps: Props, key: string): Props => {
    if (!PROPS_TO_BLOCK.includes(key)) {
      // eslint-disable-next-line no-param-reassign
      newProps[key] = props[key];
    }
    return newProps;
  }, {});
};
