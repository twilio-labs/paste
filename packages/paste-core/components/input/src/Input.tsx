import * as React from 'react';
import * as PropTypes from 'prop-types';
import {styled, css} from '@twilio-paste/styling-library';
import {InputBox} from '@twilio-paste/input-box';
import {safelySpreadFormControlProps} from './utils';

export type InputTypes = 'text' | 'email' | 'hidden' | 'number' | 'password' | 'search' | 'tel';
export type InputVariants = 'default' | 'inverse';

export interface InputProps extends React.InputHTMLAttributes<HTMLInputElement> {
  className?: never;
  disabled?: boolean;
  hasError?: boolean;
  height?: never;
  id: string;
  insertAfter?: React.ReactNode;
  insertBefore?: React.ReactNode;
  name?: string;
  placeholder?: string;
  readOnly?: boolean;
  required?: boolean;
  size?: never;
  style?: never;
  type: InputTypes;
  value: string;
  variant?: InputVariants;
  width?: never;
}

interface TypeProps {
  type: InputTypes;
  inputmode?: string | undefined;
  pattern?: string | undefined;
}

/* eslint-disable emotion/syntax-preference */
export const InputElement = styled.input<InputProps>(props =>
  css({
    appearance: 'none',
    background: 'transparent',
    border: 'none',
    borderRadius: 'borderRadius20',
    boxShadow: 'none',
    color: 'inherit',
    display: 'block',
    fontFamily: 'inherit',
    fontSize: 'fontSize30',
    fontWeight: 'fontWeightNormal',
    lineHeight: 'lineHeight20',
    outline: 'none',
    paddingBottom: 'space30',
    paddingLeft: 'space40',
    paddingRight: 'space40',
    paddingTop: 'space30',
    resize: 'none',
    width: '100%',

    '&::placeholder': {
      color: props.variant === 'inverse' ? 'colorTextInverseWeak' : 'colorTextWeak',
      fontStyle: 'italic',
    },

    '&:focus::placeholder': {
      color: props.variant === 'inverse' ? 'colorTextInverseWeak' : 'colorTextWeak',
    },

    '&:disabled': {
      color: props.variant === 'inverse' ? 'colorTextInverseWeaker' : 'colorTextWeaker',
      cursor: 'not-allowed',
    },
  })
);
/* eslint-enable */

const Input = React.forwardRef<HTMLInputElement, InputProps>(
  (
    {
      disabled,
      hasError,
      id,
      insertAfter,
      insertBefore,
      name,
      placeholder,
      readOnly,
      required,
      type,
      value,
      variant,
      ...props
    },
    ref
  ) => {
    const typeProps: TypeProps = {type};

    // https://technology.blog.gov.uk/2020/02/24/why-the-gov-uk-design-system-team-changed-the-input-type-for-numbers/
    if (type === 'number') {
      typeProps.type = 'text';
      typeProps.inputmode = 'numeric';
      typeProps.pattern = '[0-9]*';
    }

    return (
      <InputBox
        disabled={disabled}
        hasError={hasError}
        insertAfter={insertAfter}
        insertBefore={insertBefore}
        readOnly={readOnly}
        type={type}
        variant={variant}
      >
        <InputElement
          aria-invalid={hasError}
          aria-readonly={readOnly}
          {...safelySpreadFormControlProps(props)}
          {...typeProps}
          disabled={disabled}
          id={id}
          name={name}
          placeholder={placeholder}
          readOnly={readOnly}
          ref={ref}
          required={required}
          value={value}
          variant={variant}
        />
      </InputBox>
    );
  }
);

Input.displayName = 'Input';

if (process.env.NODE_ENV === 'development') {
  Input.propTypes = {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    disabled: PropTypes.bool,
    hasError: PropTypes.bool,
    id: PropTypes.string.isRequired,
    name: PropTypes.string,
    onBlur: PropTypes.func,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    placeholder: PropTypes.string,
    readOnly: PropTypes.bool,
    required: PropTypes.bool,
    type: PropTypes.oneOf(['text', 'email', 'hidden', 'number', 'password', 'search', 'tel']).isRequired as any,
    value: PropTypes.string.isRequired,
  };
}

export {Input};
