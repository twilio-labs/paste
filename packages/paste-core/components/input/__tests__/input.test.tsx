import * as React from 'react';
import {render} from '@testing-library/react';
import {axe} from 'jest-axe';
import {Theme} from '@twilio-paste/theme';
import {Label} from '@twilio-paste/label';
import {HelpText} from '@twilio-paste/help-text';
import {InputTypes, Input} from '../src';

const NOOP = (): void => {};

describe('Input render', () => {
  it('it should render', (): void => {
    const {asFragment} = render(
      <Theme.Provider theme="console">
        <Input id="input" type="text" value="test" onChange={NOOP} />
      </Theme.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('it should render with readOnly', (): void => {
    const {asFragment} = render(
      <Theme.Provider theme="console">
        <Input id="input" type="text" value="test" onChange={NOOP} readOnly />
      </Theme.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('it should render with disabled', (): void => {
    const {asFragment} = render(
      <Theme.Provider theme="console">
        <Input id="input" type="text" value="test" onChange={NOOP} disabled />
      </Theme.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('it should render with hasError', (): void => {
    const {asFragment} = render(
      <Theme.Provider theme="console">
        <Input id="input" type="text" value="test" onChange={NOOP} hasError />
      </Theme.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('it should render with prefix', (): void => {
    const {asFragment} = render(
      <Theme.Provider theme="console">
        <Input id="input" type="text" value="test" onChange={NOOP} insertBefore={<div>prefix</div>} />
      </Theme.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('it should render with suffix', (): void => {
    const {asFragment} = render(
      <Theme.Provider theme="console">
        <Input id="input" type="text" value="test" onChange={NOOP} insertAfter={<div>suffix</div>} />
      </Theme.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('it has no accessibility violations', async () => {
    const {container} = render(
      <Theme.Provider theme="console">
        <Label htmlFor="input_1">Label Text</Label>
        <Input id="input_1" type="text" value="test" onChange={NOOP} />
      </Theme.Provider>
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('disabled has no accessibility violations', async () => {
    const {container} = render(
      <Theme.Provider theme="console">
        <Label htmlFor="input_2" disabled>
          Label Text
        </Label>
        <Input id="input_2" type="text" value="test" onChange={NOOP} disabled />
      </Theme.Provider>
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  it('hasError has no accessibility violations', async () => {
    const {container} = render(
      <Theme.Provider theme="console">
        <Label htmlFor="input_3">Label Text</Label>
        <Input id="input_3" type="text" value="test" onChange={NOOP} hasError />
        <HelpText variant="error">Error info. Explains why the input has an error.</HelpText>
      </Theme.Provider>
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});

describe('Input inner input props', () => {
  const initialProps = {
    id: 'input',
    name: 'cool',
    type: 'password' as InputTypes,
    value: 'value',
    onChange: NOOP,
    disabled: true,
    readOnly: true,
    hasError: true,
    placeholder: 'Enter your name',
  };

  const {getByPlaceholderText} = render(<Input {...initialProps} />);
  const InnerInput = getByPlaceholderText(initialProps.placeholder);
  expect(InnerInput).toBeDefined();

  it('should set disabled correctly', () => {
    expect(InnerInput.getAttribute('disabled')).toEqual('');
  });

  it('should set readOnly correctly', () => {
    expect(InnerInput.getAttribute('aria-readonly')).toEqual('true');
    expect(InnerInput.getAttribute('readOnly')).toEqual('');
  });

  it('should set hasError correctly', () => {
    expect(InnerInput.getAttribute('aria-invalid')).toEqual('true');
  });

  it('should set id correctly', () => {
    expect(InnerInput.getAttribute('id')).toEqual(initialProps.id);
  });

  it('should set type correctly', () => {
    expect(InnerInput.getAttribute('type')).toEqual(initialProps.type);
  });

  it('should set value correctly', () => {
    expect(InnerInput.getAttribute('value')).toEqual(initialProps.value);
  });

  it('should set name correctly', () => {
    expect(InnerInput.getAttribute('name')).toEqual(initialProps.name);
  });

  it('should set placeholder correctly', () => {
    expect(InnerInput.getAttribute('placeholder')).toEqual(initialProps.placeholder);
  });
});

describe('Input insertBefore prop', () => {
  const initialProps = {
    id: 'input',
    type: 'text' as InputTypes,
    value: 'value',
    onChange: NOOP,
    placeholder: 'placeholder',
  };

  it('should render a prefix', () => {
    const {getByText} = render(<Input {...initialProps} insertBefore={<div>$10.99</div>} />);
    expect(getByText('$10.99')).toBeDefined();
  });

  it('should render a suffix', () => {
    const {getByText} = render(<Input {...initialProps} insertAfter={<div>$11.99</div>} />);
    expect(getByText('$11.99')).toBeDefined();
  });
});

describe('Input event handlers', () => {
  it('Should call the appropriate event handlers', () => {
    const onChangeMock: jest.Mock = jest.fn();
    const onFocusMock: jest.Mock = jest.fn();
    const onBlurMock: jest.Mock = jest.fn();

    const {getByRole} = render(
      <Input id="input" type="text" value="test" onChange={onChangeMock} onFocus={onFocusMock} onBlur={onBlurMock} />
    );
    const RenderedInput = getByRole('textbox');

    RenderedInput.focus();
    expect(onFocusMock).toHaveBeenCalledTimes(1);
    RenderedInput.blur();
    expect(onBlurMock).toHaveBeenCalledTimes(1);
  });
});

describe('Input block props', () => {
  const initialProps = {
    id: 'input',
    type: 'text' as InputTypes,
    value: 'value',
    onChange: NOOP,
    className: 'foo',
    width: '300px',
  };

  it('should not pass width and classname props', () => {
    // @ts-ignore this is on purpose
    const {getByRole} = render(<Input {...initialProps} />);
    const RenderedInput = getByRole('textbox');

    expect(RenderedInput.getAttribute('width')).toEqual(null);

    const classNames = RenderedInput.getAttribute('class');
    expect(classNames).toBeDefined();
    if (classNames != null) {
      expect(classNames.indexOf(initialProps.className)).toBe(-1);
    }
  });
});
