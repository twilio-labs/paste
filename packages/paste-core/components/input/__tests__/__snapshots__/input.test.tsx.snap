// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input render it should render 1`] = `
<DocumentFragment>
  .emotion-2 {
  color: rgb(40,42,43);
  font-size: 0.875rem;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
}

.emotion-1 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  background-color: rgb(255,255,255);
  box-shadow: 0 0 0 1px #888c8e;
  border-radius: 4px;
  -webkit-transition: box-shadow 100ms ease-in;
  transition: box-shadow 100ms ease-in;
  cursor: text;
}

.emotion-1:hover {
  box-shadow: 0 0 0 1px #003e82;
}

.emotion-1:focus-within {
  box-shadow: 0 0 0 4px rgba(2,122,197,0.8);
}

.emotion-1:active,
.emotion-1[data-active=true] {
  box-shadow: 0 0 0 1px #032c5e;
}

.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  border: none;
  border-radius: 4px;
  box-shadow: none;
  color: inherit;
  display: block;
  font-family: inherit;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.25rem;
  outline: none;
  padding-bottom: 0.5rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.5rem;
  resize: none;
  width: 100%;
}

.emotion-0::-webkit-input-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0::-moz-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0:-ms-input-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0::placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0:focus::-webkit-input-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus::-moz-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus:-ms-input-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus::placeholder {
  color: rgb(112,117,120);
}

.emotion-0:disabled {
  color: rgb(200,204,207);
  cursor: not-allowed;
}

<div
    class="emotion-2"
  >
    <div
      class="emotion-1"
      cursor="text"
      display="flex"
      width="100%"
    >
      <input
        class="emotion-0"
        id="input"
        type="text"
        value="test"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input render it should render with disabled 1`] = `
<DocumentFragment>
  .emotion-2 {
  color: rgb(40,42,43);
  font-size: 0.875rem;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
}

.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  border: none;
  border-radius: 4px;
  box-shadow: none;
  color: inherit;
  display: block;
  font-family: inherit;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.25rem;
  outline: none;
  padding-bottom: 0.5rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.5rem;
  resize: none;
  width: 100%;
}

.emotion-0::-webkit-input-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0::-moz-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0:-ms-input-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0::placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0:focus::-webkit-input-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus::-moz-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus:-ms-input-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus::placeholder {
  color: rgb(112,117,120);
}

.emotion-0:disabled {
  color: rgb(200,204,207);
  cursor: not-allowed;
}

.emotion-1 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  background-color: rgb(248,248,249);
  box-shadow: 0 0 0 1px #e4e7e9;
  border-radius: 4px;
  -webkit-transition: box-shadow 100ms ease-in;
  transition: box-shadow 100ms ease-in;
  cursor: not-allowed;
}

.emotion-1:hover {
  box-shadow: 0 0 0 1px #e4e7e9;
}

.emotion-1:focus-within {
  box-shadow: 0 0 0 4px rgba(2,122,197,0.8);
}

.emotion-1:active,
.emotion-1[data-active=true] {
  box-shadow: 0 0 0 1px #e4e7e9;
}

<div
    class="emotion-2"
  >
    <div
      class="emotion-1"
      cursor="not-allowed"
      display="flex"
      width="100%"
    >
      <input
        class="emotion-0"
        disabled=""
        id="input"
        type="text"
        value="test"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input render it should render with hasError 1`] = `
<DocumentFragment>
  .emotion-2 {
  color: rgb(40,42,43);
  font-size: 0.875rem;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
}

.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  border: none;
  border-radius: 4px;
  box-shadow: none;
  color: inherit;
  display: block;
  font-family: inherit;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.25rem;
  outline: none;
  padding-bottom: 0.5rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.5rem;
  resize: none;
  width: 100%;
}

.emotion-0::-webkit-input-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0::-moz-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0:-ms-input-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0::placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0:focus::-webkit-input-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus::-moz-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus:-ms-input-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus::placeholder {
  color: rgb(112,117,120);
}

.emotion-0:disabled {
  color: rgb(200,204,207);
  cursor: not-allowed;
}

.emotion-1 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  background-color: rgb(255,255,255);
  box-shadow: 0 0 0 1px #e22525;
  border-radius: 4px;
  -webkit-transition: box-shadow 100ms ease-in;
  transition: box-shadow 100ms ease-in;
  cursor: text;
}

.emotion-1:hover {
  box-shadow: 0 0 0 1px #921200;
}

.emotion-1:focus-within {
  box-shadow: 0 0 0 4px rgba(2,122,197,0.8);
}

.emotion-1:active,
.emotion-1[data-active=true] {
  box-shadow: 0 0 0 1px #032c5e;
}

<div
    class="emotion-2"
  >
    <div
      class="emotion-1"
      cursor="text"
      display="flex"
      width="100%"
    >
      <input
        aria-invalid="true"
        class="emotion-0"
        id="input"
        type="text"
        value="test"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input render it should render with prefix 1`] = `
<DocumentFragment>
  .emotion-3 {
  color: rgb(40,42,43);
  font-size: 0.875rem;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
}

.emotion-2 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  background-color: rgb(255,255,255);
  box-shadow: 0 0 0 1px #888c8e;
  border-radius: 4px;
  -webkit-transition: box-shadow 100ms ease-in;
  transition: box-shadow 100ms ease-in;
  cursor: text;
}

.emotion-2:hover {
  box-shadow: 0 0 0 1px #003e82;
}

.emotion-2:focus-within {
  box-shadow: 0 0 0 4px rgba(2,122,197,0.8);
}

.emotion-2:active,
.emotion-2[data-active=true] {
  box-shadow: 0 0 0 1px #032c5e;
}

.emotion-1 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  border: none;
  border-radius: 4px;
  box-shadow: none;
  color: inherit;
  display: block;
  font-family: inherit;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.25rem;
  outline: none;
  padding-bottom: 0.5rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.5rem;
  resize: none;
  width: 100%;
}

.emotion-1::-webkit-input-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-1::-moz-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-1:-ms-input-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-1::placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-1:focus::-webkit-input-placeholder {
  color: rgb(112,117,120);
}

.emotion-1:focus::-moz-placeholder {
  color: rgb(112,117,120);
}

.emotion-1:focus:-ms-input-placeholder {
  color: rgb(112,117,120);
}

.emotion-1:focus::placeholder {
  color: rgb(112,117,120);
}

.emotion-1:disabled {
  color: rgb(200,204,207);
  cursor: not-allowed;
}

.emotion-0 {
  box-sizing: border-box;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  background-color: rgb(248,248,249);
  border-bottom-left-radius: 4px;
  border-right-color: rgb(228,231,233);
  border-right-style: solid;
  border-right-width: 1px;
  border-top-left-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.25rem;
  padding: 0.5rem;
}

<div
    class="emotion-3"
  >
    <div
      class="emotion-2"
      cursor="text"
      display="flex"
      width="100%"
    >
      <div
        class="emotion-0"
        display="flex"
      >
        <div>
          prefix
        </div>
      </div>
      <input
        class="emotion-1"
        id="input"
        type="text"
        value="test"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input render it should render with readOnly 1`] = `
<DocumentFragment>
  .emotion-2 {
  color: rgb(40,42,43);
  font-size: 0.875rem;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
}

.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  border: none;
  border-radius: 4px;
  box-shadow: none;
  color: inherit;
  display: block;
  font-family: inherit;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.25rem;
  outline: none;
  padding-bottom: 0.5rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.5rem;
  resize: none;
  width: 100%;
}

.emotion-0::-webkit-input-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0::-moz-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0:-ms-input-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0::placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0:focus::-webkit-input-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus::-moz-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus:-ms-input-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus::placeholder {
  color: rgb(112,117,120);
}

.emotion-0:disabled {
  color: rgb(200,204,207);
  cursor: not-allowed;
}

.emotion-1 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  background-color: rgb(248,248,249);
  box-shadow: 0 0 0 1px #888c8e;
  border-radius: 4px;
  -webkit-transition: box-shadow 100ms ease-in;
  transition: box-shadow 100ms ease-in;
  cursor: text;
}

.emotion-1:hover {
  box-shadow: 0 0 0 1px #888c8e;
}

.emotion-1:focus-within {
  box-shadow: 0 0 0 4px rgba(2,122,197,0.8);
}

.emotion-1:active,
.emotion-1[data-active=true] {
  box-shadow: 0 0 0 1px #032c5e;
}

<div
    class="emotion-2"
  >
    <div
      class="emotion-1"
      cursor="text"
      display="flex"
      width="100%"
    >
      <input
        aria-readonly="true"
        class="emotion-0"
        id="input"
        readonly=""
        type="text"
        value="test"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Input render it should render with suffix 1`] = `
<DocumentFragment>
  .emotion-3 {
  color: rgb(40,42,43);
  font-size: 0.875rem;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.25rem;
  font-weight: 400;
}

.emotion-2 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  background-color: rgb(255,255,255);
  box-shadow: 0 0 0 1px #888c8e;
  border-radius: 4px;
  -webkit-transition: box-shadow 100ms ease-in;
  transition: box-shadow 100ms ease-in;
  cursor: text;
}

.emotion-2:hover {
  box-shadow: 0 0 0 1px #003e82;
}

.emotion-2:focus-within {
  box-shadow: 0 0 0 4px rgba(2,122,197,0.8);
}

.emotion-2:active,
.emotion-2[data-active=true] {
  box-shadow: 0 0 0 1px #032c5e;
}

.emotion-0 {
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: transparent;
  border: none;
  border-radius: 4px;
  box-shadow: none;
  color: inherit;
  display: block;
  font-family: inherit;
  font-size: 0.875rem;
  font-weight: 400;
  line-height: 1.25rem;
  outline: none;
  padding-bottom: 0.5rem;
  padding-left: 0.75rem;
  padding-right: 0.75rem;
  padding-top: 0.5rem;
  resize: none;
  width: 100%;
}

.emotion-0::-webkit-input-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0::-moz-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0:-ms-input-placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0::placeholder {
  color: rgb(112,117,120);
  font-style: italic;
}

.emotion-0:focus::-webkit-input-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus::-moz-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus:-ms-input-placeholder {
  color: rgb(112,117,120);
}

.emotion-0:focus::placeholder {
  color: rgb(112,117,120);
}

.emotion-0:disabled {
  color: rgb(200,204,207);
  cursor: not-allowed;
}

.emotion-1 {
  box-sizing: border-box;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  background-color: rgb(248,248,249);
  border-bottom-right-radius: 4px;
  border-left-color: rgb(228,231,233);
  border-left-style: solid;
  border-left-width: 1px;
  border-top-right-radius: 4px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  line-height: 1.25rem;
  padding: 0.5rem;
}

<div
    class="emotion-3"
  >
    <div
      class="emotion-2"
      cursor="text"
      display="flex"
      width="100%"
    >
      <input
        class="emotion-0"
        id="input"
        type="text"
        value="test"
      />
      <div
        class="emotion-1"
        display="flex"
      >
        <div>
          suffix
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
